core/lib/src/catcher/catcher.rs

```diff
@@ -127,6 +134,9 @@ pub struct Catcher {
     ///
     /// This is -(number of nonempty segments in base).
     pub(crate) rank: isize,
+
+    /// A unique route type to identify this route
+    pub(crate) catcher_type: Option<TypeId>,
```

Comment:
Very cool! I had to read about [TypeId](https://doc.rust-lang.org/std/any/struct.TypeId.html).